bin/zlock

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python
+#
+# zlock - lock a ZFS filesystem from being replicated by zbackup.
+#
+# In fact, the parameter filesystem is not validated to be a real ZFS filesystem,
+# since we choose not to care if the user locks a non-existent filesystem.
+#
+# There is no validation that the unlocker is the one holding the lock.
+#
+# Author: Simon Guest, 22/9/2014
+# Licensed under GNU General Public License GPLv3
+
+import optparse
+import os
+import os.path
+from stat import ST_MTIME
+import sys
+import time
+from zfstools.util import stderr, verbose_stderr, set_verbose
+
+LOCKDIR = "/var/lib/zfs-tools/zlock"
+
+def ensure_lockdir_exists():
+    if not os.path.isdir(LOCKDIR):
+        os.makedirs(LOCKDIR, 0755)
+
+def filesystem2lockfile(filesystem):
+    return filesystem.replace('/', ';')
+
+def lockfile2filesystem(lockfile):
+    return lockfile.replace(';', '/')
+
+def readme_comment(readme, prefix):
+    try:
+        with open(readme, 'r') as f:
+            comment = "%s%s" % (prefix, f.read().rstrip('\n'))
+    except IOError:
+        comment = ""
+    return comment
+
+def lockpath_comment(lockpath, prefix):
+    return readme_comment(os.path.join(lockpath, "README"), prefix)
+
+def lock(filesystem, options):
+    """Attempt to lock filesystem, return whether we did."""
+    lockpath = os.path.join(LOCKDIR, filesystem2lockfile(filesystem))
+    readme = os.path.join(lockpath, "README")
+    if options.dryrun:
+        # test whether lock is available, don't actually acquire it
+        if os.path.isdir(lockpath):
+            verbose_stderr("zlock: (dry-run) would fail to lock %s%s" % (filesystem, readme_comment(readme, " - ")))
+            return False
+        else:
+            verbose_stderr("zlock: (dry-run) would lock %s" % filesystem)
+            return True
+    else:
+        try:
+            os.mkdir(lockpath)
```

Comment:
Race condition.  Two processes could hit this twice, one would bomb with EEXIST.  Protect against EEXIST with an exception handler.